newrelic/core/node_mixin.py

```diff
@@ -79,71 +79,91 @@ def span_event(
+
+        if ct_exit_spans is None:
+            ct_exit_spans = {"instrumented": 0, "kept": 0, "dropped_ids": 0}
+        ct_exit_spans["instrumented"] += 1
+
+        partial_granularity_type = settings.distributed_tracing.sampler.partial_granularity.type
+        a_attrs_set = set(a_attrs)
+        exit_span_attrs_present = attribute.SPAN_ENTITY_RELATIONSHIP_ATTRIBUTES & a_attrs_set
+        exit_span_error_attrs_present = attribute.SPAN_ERROR_ATTRIBUTES & a_attrs_set
+        # If this is an entry span, add `nr.pg` to indicate transaction is partial
+        # granularity sampled.
+        if i_attrs.get("nr.entryPoint"):
+            a_attrs["nr.pg"] = True
+        # If this is the entry node or an LLM span always return it.
+        if i_attrs.get("nr.entryPoint") or i_attrs["name"].startswith("Llm/"):
+            ct_exit_spans["kept"] += 1
             if partial_granularity_type == "reduced":
                 return [i_attrs, u_attrs, a_attrs]
             else:
-                a_minimized_attrs = attr_class({key: a_attrs[key] for key in exit_span_attrs_present})
-                # If we are in essential mode return the span with minimized attributes.
-                if partial_granularity_type == "essential":
-                    return [i_attrs, {}, a_minimized_attrs]
-                # If the span is an exit span but span compression (compact) is enabled,
-                # we need to check for uniqueness before returning it.
-                # Combine all the entity relationship attr values into a string to be
-                # used as the hash to check for uniqueness.
-                span_attrs = "".join([str(a_minimized_attrs[key]) for key in exit_span_attrs_present])
-                new_exit_span = span_attrs not in ct_exit_spans
-                # If this is a new exit span, add it to the known ct_exit_spans and
-                # return it.
-                if new_exit_span:
-                    # nr.ids is the list of span guids that share this unqiue exit span.
-                    a_minimized_attrs["nr.ids"] = []
-                    a_minimized_attrs["nr.durations"] = self.duration
-                    ct_exit_spans[span_attrs] = [i_attrs, a_minimized_attrs]
-                    return [i_attrs, {}, a_minimized_attrs]
-                # If this is an exit span we've already seen, add it's guid to the list
-                # of ids on the seen span, compute the new duration & start time, and
-                # return None.
+                return [i_attrs, {}, {key: a_attrs.get(key) for key in exit_span_error_attrs_present | {"nr.pg"}}]
+        # If the span is not an exit span, skip it by returning None.
+        if not exit_span_attrs_present:
+            return None
+        # If the span is an exit span and we are in reduced mode (meaning no attribute dropping),
+        # just return the exit span as is.
+        if partial_granularity_type == "reduced":
+            ct_exit_spans["kept"] += 1
```

Comment:
There was a bug here that I fixed which is why this is here now. I had to move these into here because gets incremented multiple times if it's put in span_events because span_events loops over the children and then the children of the children (so it gets incremented multiple times each time is yields the node up the chain) where as here it is guaranteed that it is only incremented once.